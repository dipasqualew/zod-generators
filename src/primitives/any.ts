import type { ZodAny, ZodUnknown } from 'zod';

/**
 * Generates valid any/unknown mocks
 * from the given ZodAny/ZodUnknown definition
 *
 * @param _field
 */
export const mockValid = (_field: ZodAny | ZodUnknown) => {
  return {
    DEFAULT: 1,
  };
};

/**
 * Generates invalid any/unknown mocks
 * from the given ZodAny/ZodUnknown definition
 *
 * @param _field
 */
export const mockInvalid = (_field: ZodAny | ZodUnknown) => {
  // Any and Unknown are always valid
  return {};
};
