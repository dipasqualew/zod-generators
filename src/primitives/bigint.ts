import type { ZodBigInt } from 'zod';

import { MAX_INTEGER, MIN_INTEGER, getRandomNumber } from '../utils';

/**
 * Generates valid bigint mocks
 * from the given ZodBigInt definition
 *
 * @param _field
 */
export const mockValid = (_field: ZodBigInt) => {
  const min = MIN_INTEGER;
  const max = MAX_INTEGER;

  const numbers = {
    DEFAULT: BigInt(getRandomNumber(min, max, true)),
    MIN: BigInt(min),
    MAX: BigInt(max),
  };

  return numbers;
};

/**
 * Generates invalid bigint mocks
 * from the given ZodBigInt definition
 *
 * @param _field
 */
export const mockInvalid = (_field: ZodBigInt) => {
  const min = MIN_INTEGER;
  const max = MAX_INTEGER;

  return {
    DEFAULT: 'not-a-number',
    NUMBER: getRandomNumber(min, max, true),
    FLOAT: getRandomNumber(min, max, true) + 0.1,
  };
};
