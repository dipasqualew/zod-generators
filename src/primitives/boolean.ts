import type { ZodBoolean } from 'zod';

/**
 * Generates valid boolean mocks
 * from the given ZodBoolean definition
 *
 * @param _field
 */
export const mockValid = (_field: ZodBoolean) => {
  const booleans = {
    DEFAULT: true,
    TRUE: true,
    FALSE: false,
  };

  return booleans;
};

/**
 * Generates invalid boolean mocks
 * from the given ZodBoolean definition
 *
 * @param _field
 */
export const mockInvalid = (_field: ZodBoolean) => {
  const booleans = {
    DEFAULT: NaN,
  };

  return booleans;
};
