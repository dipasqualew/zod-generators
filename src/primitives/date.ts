import type { ZodDate } from 'zod';

/**
 * Generates valid boolean mocks
 * from the given ZodDate definition
 *
 * @param _field
 */
export const mockValid = (_field: ZodDate) => {
  const booleans = {
    DEFAULT: new Date(),
  };

  return booleans;
};

/**
 * Generates invalid boolean mocks
 * from the given ZodDate definition
 *
 * @param _field
 */
export const mockInvalid = (_field: ZodDate) => {
  const booleans = {
    DEFAULT: 'not-a-date',
    INVALID_DATE: new Date('not-a-date'),
  };

  return booleans;
};
