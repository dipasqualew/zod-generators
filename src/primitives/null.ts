import type { ZodNull } from 'zod';

/**
 * Generates valid null mocks
 * from the given ZodNull definition
 *
 * @param _field
 */
export const mockValid = (_field: ZodNull) => {
  return {
    DEFAULT: null,
  };
};

/**
 * Generates invalid any/unknown mocks
 * from the given ZodNull definition
 *
 * @param _field
 */
export const mockInvalid = (_field: ZodNull) => {
  return {
    DEFAULT: 'DEFAULT',
    STRING: 'STRING',
    NUMBER: 1,
  };
};
