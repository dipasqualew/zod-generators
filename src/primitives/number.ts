import type { ZodNumber } from 'zod';

import { MAX_INTEGER, MIN_INTEGER, getRandomNumber } from '../utils';

/**
 * Generates valid number mocks
 * from the given ZodNumber definition
 *
 * @param field
 */
export const mockValid = (field: ZodNumber) => {
  const min = field._def.minimum?.value || MIN_INTEGER;
  const max = field._def.maximum?.value || MAX_INTEGER;
  const integer = Boolean(field._def.isInteger);

  const numbers = {
    DEFAULT: getRandomNumber(min, max, integer),
    MIN: min,
    MAX: max,
  };

  return numbers;
};

/**
 * Generates invalid number mocks
 * from the given ZodNumber definition
 *
 * @param field
 */
export const mockInvalid = (field: ZodNumber) => {
  const integer = Boolean(field._def.isInteger);

  const strings: [string, string | number][] = [
    ['DEFAULT', 'not-a-number'],
  ];

  if (field._def.minimum) {
    const underMin = getRandomNumber(MIN_INTEGER, field._def.minimum.value - 1);
    strings.push(['MIN', underMin]);
  }

  if (field._def.maximum) {
    const overMax = getRandomNumber(field._def.maximum.value + 1, MAX_INTEGER);
    strings.push(['MAX', overMax]);
  }

  if (integer) {
    const float = getRandomNumber(field._def.minimum?.value || 0, field._def.maximum?.value || 20, true) + 0.1;
    strings.push(['FLOAT', float]);
  }

  return Object.fromEntries(strings) as { DEFAULT: number } & Record<string, number>;
};
