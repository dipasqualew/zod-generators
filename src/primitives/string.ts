import { v4 as uuid4 } from 'uuid';
import type { ZodString } from 'zod';

import {
  getRandomEmail,
  getRandomString,
  getRandomUrl,
  getString,
} from '../utils';

/**
 * Generates valid string mocks
 * from the given ZodString definition
 *
 * @param field
 */
export const mockValid = (field: ZodString) => {
  const minLength = field._def.minLength?.value || 0;
  const maxLength = field._def.maxLength?.value || (minLength + 64);

  let generator = (min: number, max: number) => getRandomString(min, max);

  if (field._def.isUUID) {
    // UUIDs are fix length
    // return immediately
    return {
      DEFAULT: uuid4(),
    };
  }

  if (field._def.isEmail) {
    generator = (min: number, max: number) => getRandomEmail(min, max);
  }

  if (field._def.isURL) {
    generator = (min: number, max: number) => getRandomUrl(min, max);
  }

  const strings = {
    DEFAULT: generator(minLength, maxLength),
    MIN: generator(minLength, maxLength),
    MAX: generator(maxLength, maxLength),
  };

  return strings;
};

/**
 * Generates invalid string mocks
 * from the given ZodString definition
 *
 * @param field
 */
export const mockInvalid = (field: ZodString) => {
  const minLength = field._def.minLength?.value || 0;

  const strings: [string, string | null][] = [
    ['DEFAULT', null],
  ];

  if (minLength > 0) {
    strings.push(['MIN', getString(minLength - 1)]);
  }

  if (field._def.maxLength) {
    strings.push(['MAX', getString(field._def.maxLength.value + 1)]);
  }

  return Object.fromEntries(strings) as { DEFAULT: string } & Record<string, string>;
};
