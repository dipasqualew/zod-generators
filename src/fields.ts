/* eslint-disable no-use-before-define */
import * as zod from 'zod';

import * as literal from './literal';
import * as any from './primitives/any';
import * as bigint from './primitives/bigint';
import * as boolean from './primitives/boolean';
import * as date from './primitives/date';
import * as znull from './primitives/null';
import * as number from './primitives/number';
import * as string from './primitives/string';
import * as zundefined from './primitives/undefined';

/**
 * Mock generator function
 */
export type MockGenerator = () => { DEFAULT?: unknown } & Record<string, unknown>;

/**
 * Returns the generator of valid values
 * for the given field
 *
 * @param field
 */
export const getValidGenerator = <T extends zod.ZodAny>(field: T): MockGenerator => {
  if (field instanceof zod.ZodString) {
    return () => string.mockValid(field);
  }
  if (field instanceof zod.ZodNumber) {
    return () => number.mockValid(field);
  }

  if (field instanceof zod.ZodBoolean) {
    return () => boolean.mockValid(field);
  }

  if (field instanceof zod.ZodArray) {
    return () => generateValidArrays(field);
  }

  if (field instanceof zod.ZodObject) {
    return () => generateValidObjects(field as zod.ZodObject<zod.ZodRawShape>);
  }

  if (field instanceof zod.ZodDate) {
    return () => date.mockValid(field);
  }

  if (field instanceof zod.ZodBigInt) {
    return () => bigint.mockValid(field);
  }

  if (field instanceof zod.ZodLiteral) {
    return () => literal.mockValid(field);
  }

  if (field instanceof zod.ZodNull) {
    return () => znull.mockValid(field);
  }

  if (field instanceof zod.ZodUndefined || field instanceof zod.ZodVoid) {
    return () => zundefined.mockValid(field);
  }

  return () => any.mockValid(field);
};

/**
 * Returns the generator of invalid values
 * for the given field
 *
 * @param field
 */
export const getInvalidGenerator = <T extends zod.ZodAny>(field: T): MockGenerator => {
  if (field instanceof zod.ZodString) {
    return () => string.mockInvalid(field);
  }

  if (field instanceof zod.ZodNumber) {
    return () => number.mockInvalid(field);
  }

  if (field instanceof zod.ZodBoolean) {
    return () => boolean.mockInvalid(field);
  }

  if (field instanceof zod.ZodArray) {
    return () => generateInvalidArrays(field);
  }

  if (field instanceof zod.ZodObject) {
    return () => generateInvalidObjects(field as zod.ZodObject<zod.ZodRawShape>);
  }

  if (field instanceof zod.ZodDate) {
    return () => date.mockInvalid(field);
  }

  if (field instanceof zod.ZodLiteral) {
    return () => literal.mockInvalid(field);
  }

  if (field instanceof zod.ZodBigInt) {
    return () => bigint.mockInvalid(field);
  }

  if (field instanceof zod.ZodNull) {
    return () => znull.mockInvalid(field);
  }

  if (field instanceof zod.ZodUndefined || field instanceof zod.ZodVoid) {
    return () => zundefined.mockInvalid(field);
  }

  return () => any.mockInvalid(field);
};

/**
 * Returns the valid mocks
 * for the given field
 *
 * @param field
 */
export const mockValid = <T extends zod.ZodAny>(field: T) => {
  const generator = getValidGenerator(field);

  return generator();
};

/**
 * Returns the invalid mocks
 * for the given field
 *
 * @param field
 */
export const mockInvalid = <T extends zod.ZodAny>(field: T) => {
  const generator = getInvalidGenerator(field);

  return generator();
};

/**
 * Returns all mocks
 * for the given field
 *
 * @param field
 */
export const mock = <T extends zod.ZodAny>(field: T) => {
  return {
    valid: mockValid(field),
    invalid: mockInvalid(field),
  };
};

/**
 * Generates valid arrays
 * from the given ZodArray definition
 *
 * @param field
 */
export const generateValidArrays = <T extends zod.ZodAny>(field: zod.ZodArray<T>) => {
  const generator = getValidGenerator(field._def.type);
  const items = generator();

  return {
    DEFAULT: Object.values(items),
  };
};

/**
 * Generates invalid arrays
 * from the given ZodArray definition
 *
 * @param field
 */
export const generateInvalidArrays = <T extends zod.ZodAny>(field: zod.ZodArray<T>) => {
  const generator = getInvalidGenerator(field._def.type);
  const items = generator();

  const values = Object.values(items);

  if (values.length) {
    return {
      DEFAULT: values,
    };
  }

  return {};
};

/**
 * Generates valid objects
 * from the given ZodObject definition
 *
 * @param field
 */
export const generateValidObjects = <T extends zod.ZodRawShape>(field: zod.ZodObject<T>) => {
  type Shape = zod.infer<typeof field>;

  const defaultMock: Partial<Shape> = {};
  const variations: Record<string, Record<string, unknown>> = {};

  Object.entries(field.shape).forEach((entry) => {
    const key = entry[0] as keyof Shape;
    const def = entry[1] as Shape[keyof Shape];
    const generator = getValidGenerator(def);
    const validMocks = generator();

    defaultMock[key] = validMocks.DEFAULT as Shape[keyof Shape];
    variations[entry[0]] = validMocks;
  });

  const mocks = [
    ['DEFAULT', defaultMock],
  ];

  Object.entries(variations).forEach((variation) => {
    Object.entries(variation[1]).forEach((entry) => {
      const name = `${variation[0]}__${entry[0]}`;
      const value = { ...defaultMock, [variation[0]]: entry[1] };

      mocks.push([name, value]);
    });
  });

  return Object.fromEntries(mocks) as { DEFAULT: Shape } & Record<string, Shape>;
};

/**
 * Generates valid objects
 * from the given ZodObject definition
 *
 * @param field
 */
export const generateInvalidObjects = <T extends zod.ZodRawShape>(field: zod.ZodObject<T>) => {
  type Shape = zod.infer<typeof field>;

  const valid = generateValidObjects(field).DEFAULT;

  const defaultMock: Partial<Shape> = {};
  const variations: Record<string, Record<string, unknown>> = {};

  Object.entries(field.shape).forEach((entry) => {
    const key = entry[0] as keyof Shape;
    const def = entry[1] as Shape[keyof Shape];
    const generator = getInvalidGenerator(def);
    const invalidMocks = generator();

    if ('DEFAULT' in invalidMocks) {
      defaultMock[key] = invalidMocks.DEFAULT as Shape[keyof Shape];
      variations[entry[0]] = invalidMocks;
    }
  });

  const mocks: [string, unknown][] = [];

  if (Object.keys(defaultMock).length) {
    mocks.push(['DEFAULT', defaultMock]);
  }

  Object.entries(variations).forEach(([fieldName, fieldVariations]) => {
    Object.entries(fieldVariations).forEach(([description, fieldValue]) => {
      const name = `${fieldName}__${description}`;
      const value = { ...valid, [fieldName]: fieldValue };

      mocks.push([name, value]);
    });
  });

  return Object.fromEntries(mocks) as { DEFAULT: unknown } & Record<string, unknown>;
};
