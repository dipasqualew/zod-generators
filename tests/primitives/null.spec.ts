import * as znull from '../../src/primitives/null';
import { nullFields } from '../fields';
import { invalidateGroup, validateGroup } from '../mixins';

describe('null', () => {
  nullFields.forEach(([description, field]) => {
    describe(description, () => {
      describe('validates', () => {
        const valid = znull.mockValid(field);
        validateGroup(field, valid);
      });

      describe('invalidates', () => {
        const valid = znull.mockInvalid(field);
        invalidateGroup(field, valid);
      });
    });
  });
});
