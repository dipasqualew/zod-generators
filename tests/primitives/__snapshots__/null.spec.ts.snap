// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`null a null field invalidates Invalidates DEFAULT 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"null\\",
    \\"received\\": \\"string\\",
    \\"path\\": [],
    \\"message\\": \\"Expected null, received string\\"
  }
]"
`;

exports[`null a null field invalidates Invalidates NUMBER 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"null\\",
    \\"received\\": \\"number\\",
    \\"path\\": [],
    \\"message\\": \\"Expected null, received number\\"
  }
]"
`;

exports[`null a null field invalidates Invalidates STRING 1`] = `
"[
  {
    \\"code\\": \\"invalid_type\\",
    \\"expected\\": \\"null\\",
    \\"received\\": \\"string\\",
    \\"path\\": [],
    \\"message\\": \\"Expected null, received string\\"
  }
]"
`;
